mm2src/coins/eth.rs

```diff
@@ -3403,14 +3404,33 @@ impl EthCoin {
                 let mut value = U256::from(0);
                 let mut amount = trade_amount;
 
+                debug!("Using watcher reward {:?} for swap payment", args.watcher_reward);
+
                 let data = match args.watcher_reward {
                     Some(reward) => {
-                        let reward_amount = try_tx_fus!(wei_from_big_decimal(&reward.amount, self.decimals));
-
-                        match reward.reward_target {
-                            RewardTarget::Contract | RewardTarget::PaymentSender => value += reward_amount,
-                            RewardTarget::PaymentSpender => amount += reward_amount,
-                            _ => (),
+                        let reward_amount = match reward.reward_target {
+                            RewardTarget::Contract | RewardTarget::PaymentSender => {
+                                let eth_reward_amount = try_tx_fus!(wei_from_big_decimal(&reward.amount, ETH_DECIMALS));
+                                value += eth_reward_amount;
+                                eth_reward_amount
+                            },
+                            RewardTarget::PaymentSpender => {
+                                let token_reward_amount =
+                                    try_tx_fus!(wei_from_big_decimal(&reward.amount, self.decimals));
+                                amount += token_reward_amount;
+                                token_reward_amount
+                            },
+                            _ => {
+                                // TODO tests passed without this change, need to research on how it worked
+                                if reward.send_contract_reward_on_spend {
+                                    let eth_reward_amount =
+                                        try_tx_fus!(wei_from_big_decimal(&reward.amount, ETH_DECIMALS));
+                                    value += eth_reward_amount;
+                                    eth_reward_amount
+                                } else {
+                                    0.into()
+                                }
```

Comment:
IIRC, all watcher tests were passing with old testnet environment but started to fail after switching to dockerized Geth. `send_contract_reward_on_spend` can be true in `get_maker_watcher_reward` if `other_coin.is_eth()`:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/8635ed94a7c3334bce94c4f6d5fabbf5f227e92c/mm2src/coins/eth.rs#L2052

And when it's true, the contract attempts to send additional ETH, which isn't available on its balance. With this change, the reward ETH is also transferred to a contract during payment, making ETH/ERC20 watcher tests pass.

> P.S. I think watchers code needs a lot of refactors to separate it from default swap case logic.

I agree with this. My goal was to make tests green and learn ETH watchers code/understand it better. It seems that `RewardTarget` and `send_contract_reward_on_spend` can be removed to handle most of the logic on smart contract side. I would like to either work on it myself in the future or guide the developer who will take the task.

Reply:
> send_contract_reward_on_spend can be true in get_maker_watcher_reward if other_coin.is_eth()

Yeah, missed this since I was looking for cases with `RewardTarget::PaymentReceive` as looking for `None` case is not logical.

> I would like to either work on it myself in the future or guide the developer who will take the task.

Sure. Thanks a lot for offering your help on this :)